core/src/main/java/org/apache/carbondata/core/scan/expression/geo/PolygonExpression.java

```diff
@@ -0,0 +1,118 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.scan.expression.geo;
+
+import java.util.List;
+
+import org.apache.carbondata.common.annotations.InterfaceAudience;
+import org.apache.carbondata.core.metadata.datatype.DataTypes;
+import org.apache.carbondata.core.scan.expression.ColumnExpression;
+import org.apache.carbondata.core.scan.expression.Expression;
+import org.apache.carbondata.core.scan.expression.ExpressionResult;
+import org.apache.carbondata.core.scan.expression.LiteralExpression;
+import org.apache.carbondata.core.scan.expression.conditional.GreaterThanEqualToExpression;
+import org.apache.carbondata.core.scan.expression.conditional.LessThanEqualToExpression;
+import org.apache.carbondata.core.scan.expression.exception.FilterIllegalMemberException;
+import org.apache.carbondata.core.scan.expression.exception.FilterUnsupportedException;
+import org.apache.carbondata.core.scan.expression.logical.AndExpression;
+import org.apache.carbondata.core.scan.expression.logical.OrExpression;
+import org.apache.carbondata.core.scan.expression.logical.RangeExpression;
+import org.apache.carbondata.core.scan.expression.logical.TrueExpression;
+import org.apache.carbondata.core.scan.filter.intf.ExpressionType;
+import org.apache.carbondata.core.scan.filter.intf.RowIntf;
+import org.apache.carbondata.core.util.CustomIndex;
+
+/**
+ * InPolygon expression processor. It inputs the InPolygon string to the GeoHash implementation's
+ * query method, gets the list of ranges of GeoHash IDs to filter as an output. And then, multiple
+ * range expressions are build from those list of ranges.
+ */
+@InterfaceAudience.Internal
+public class PolygonExpression extends Expression {
+  private String polygon;
+  private String columnName;
+  private CustomIndex<Long, String, List<Long[]>> handler;
+  private List<Long[]> ranges;
+
+  public PolygonExpression(String polygon, String columnName, CustomIndex handler) {
+    this.polygon = polygon;
+    this.handler = handler;
+    this.columnName = columnName;
+  }
+
+  /**
+   * This method builds the GeoHash range expressions from the list of ranges of GeoHash IDs.
+   */
+  public void buildRangeExpression() {
+    try {
+      ranges = handler.query(polygon);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+    // Convert these ranges into range expressions
+    Expression expression = null;
+    Expression prevExpression = null;
+    Expression rangeExpression;
+    for (Long[] range : ranges) {
+      assert (range.length == 2);
```

Comment:
we get quadtree results like 1-3, 4, 5, 6-9 ?

so in this case range length != 2 ? Also in above example, 4, 5 need to go as EqualToExpression ?

Reply:
@VenuReddy2103 And in case if range size is one then add also need to add One EqualToExpression

Reply:
Yes, Modified to EqualToExpression now.